torch/utils/_sympy/functions.py

```diff
@@ -589,8 +603,23 @@ def __new__(cls, *args, **assumptions):
         # base creation
         obj = Expr.__new__(cls, *ordered(args), **assumptions)
         obj._argset = args
+
         return obj
 
+    @classmethod
+    def _unique_atoms_no_constants(cls, args):
+        """Return true if all atoms in all args are all unique and there are no constants atoms"""
+        seen_symbols = set()
+        for arg in args:
+            for element in arg.atoms():
```

Comment:
The use of atoms here seems suspicious. Why is it ok to iterate over atoms?

Reply:
i can restrict this to summations of symbols that are unique, but my intuition was the following:

iterating over atoms above implies the following:
1. There is no constants in any input expression.
2. There are no shared symbols across expressions. 

that applies (no proof here)
1. We cant tell that some arg is bigger or smaller than another (well because they are made of complete different symbols and there is no numeric values on each side). 
2. We can have redundant expressions or common elements (same reasoning above) 

so I could not think of max(e1, e2)
where e1 and e2 do not have numbers, do not share symbols but i can tell that one is greater than the other. 

Finally those opts sounded like best effort anyway from the comment above (remove redundant args that are easily identified). so i thought in the worse case if i missed one, its clearly a complicated one !

let me know if that convince you , that said I am happy to restrict this to summations of symbols of form (Max(a+b, c+d, ..)).

Reply:
One more thing is its also assumed inputs are already simplified , none of them is (a-a) or (b-b) which is valid assumtion.

Reply:
It is difficult to see without an example.

It seems I can generate a synthetic function that will cause you to go wrong. Let MulByZero be a function which multiplies by zero. We now have Min(MulByZero(s0), MulByZero(s1)). The atoms of MulByZero(s0) are simply s0. By the logic here, there are no non-symbol atoms and therefore everything is unique. But clearly I can simplify this whole shebang to zero.

In general, when you reason about atoms, you are saying that whenever I have Min(F1(s0, ...), F2(s1, ...)), *NO MATTER* the choice of F1/F2, there is guaranteed not to be any possible simplification from collapse arguments / localzeros. This seems difficult to prove!

Reply:
I mean MulByZero is synthetic function, has it been x*0, it would have been 0, since inputs are already simplified .

I understand that this is difficult to prove and in fact not correct for all type of expressions that anyone can create, but it sounds to me that it would work with the type of expressions that we deal with in pytorch here. run_extra_validations will catch if we ran anything in Pytorch that would invalidate that (in the tests).

That said, since in this benchmark we are dealing with summations inputs. I will just rewrite it to make stricter to that.

Reply:
So, just to recap discussion from offsite, we were *only* going to look into summations (and potentially polynomials too), because that's what you actually need in practice and otherwise it's hard to reason through what exactly the optimization here is.

Once you limit to summations, you can just say: "Let's suppose we are doing Max(polynomial1, polynomial2)". When the symbols of polynomial1 and polynomial2 are distinct, we know it's not possible for there to be any potential simplification to the formula here, so we bypass simplification."

By the way, a *better* (i.e., more permanent) approach is to take the overly general preexisting simplification logic in Sympy and figure out what part of it actually matters for ML workloads, and then rewrite all of it to just do that simple core (which is also fast!) This is a bit harder to do though because you have to figure out what actually matters and what doesn't, so there may be some trial and error.